README.rst
odoo-elasticapm
=================

.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.png
    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
    :alt: License: AGPL-3
.. image:: https://badge.fury.io/py/odoo-elasticapm.svg
    :target: http://badge.fury.io/py/odoo-elasticapm

odoo-elasticapm integrate the APM tracking from the ELK suite

.. contents::

Quick start
~~~~~~~~~~~

Install ``odoo-elasticapm``::

  pip install odoo-elasticapm


Then instead of launching odoo with ``odoo`` cmd use ``odoo-elasticapm``::

  odoo-elasticapm


Configuration

Following environment variable are needed:

ELASTIC_APM_SERVER_URL=http://apm-server:8200


The following one are optionnal:

ELASTIC_APM_SERVICE_NAME=my-customer
ELASTIC_APM_TRANSACTION_SAMPLE_RATE=0.1

All environment variable are available on official documentation:
https://www.elastic.co/guide/en/apm/agent/python/current/configuration.html


Issue
~~~~~~~
We still have issue on travis regarding verison 12 and 13 (I fail to reproduce them locally)
If you want to use it on this version be carefull.
Contibution and help is welcome


CHANGES.rst
Changes
~~~~~~~

.. Future (?)
.. ----------
.. - ...


0.0.1
--------

- add transaction capture for http request
- add transaction capture for cron
- add exception capture for http request
- add exception capture for cron
- add special span for ORM method (search, create, write, unlink)


odoo_elasticapm/base.py
# -*- coding: utf-8 -*-


import os

try:
    from odoo.tools.config import config
    import odoo
except ImportError:
    from openerp.tools.config import config
    import openerp as odoo

# The elasticapm lib must be imported just after odoo
# so the odoo.evented variable will be correctly defined

import elasticapm

odoo_version = odoo.release.version


def version_older_then(version):
    return (
        odoo.tools.parse_version(odoo_version)[0] < odoo.tools.parse_version(version)[0]
    )


if os.environ.get("ELASTIC_APM_ENVIRONMENT"):
    environment = os.environ.get("ELASTIC_APM_ENVIRONMENT")
else:
    environment = config.get("running_env")

elasticapm.instrument()

elastic_apm_client = elasticapm.Client(
    framework_name="Odoo",
    framework_version=odoo_version,
    service_name=os.environ.get("ELASTIC_APM_SERVICE_NAME", "Odoo"),
    environment=environment,
)


odoo_elasticapm/http.py
# -*- coding: utf-8 -*-
# Copyright 2019 Akretion (http://www.akretion.com).
# @author Sébastien BEAU <sebastien.beau@akretion.com>
# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).

from .base import elastic_apm_client, elasticapm

try:
    from odoo.http import WebRequest, request
    from odoo.exceptions import (
        UserError,
        RedirectWarning,
        AccessDenied,
        AccessError,
        MissingError,
        ValidationError,
        except_orm,
    )


except ImportError:
    from openerp.http import WebRequest, request
    from openerp.exceptions import (
        Warning as UserError,
        RedirectWarning,
        AccessDenied,
        AccessError,
        MissingError,
        ValidationError,
        except_orm,
    )

EXCEPTIONS = [
    UserError,
    RedirectWarning,
    AccessDenied,
    AccessError,
    MissingError,
    ValidationError,
    except_orm,
]


def get_data_from_request():
    httprequest = request.httprequest
    data = {
        "headers": dict(**httprequest.headers),
        "method": httprequest.method,
        "socket": {
            "remote_address": httprequest.remote_addr,
            "encrypted": httprequest.scheme == "https",
        },
        "url": elasticapm.utils.get_url_dict(httprequest.url),
    }
    # remove Cookie header since the same data is in request["cookies"] as well
    data["headers"].pop("Cookie", None)
    return data


ori_handle_exception = WebRequest._handle_exception


def _handle_exception(self, exception):
    handled = False
    for exception_class in EXCEPTIONS:
        if isinstance(exception, exception_class):
            handled = True
    elasticapm.label(
        exception_source="request",
        exception_type=type(exception).__name__,
        exception_handled=handled,
    )
    elastic_apm_client.capture_exception(
        context={"request": get_data_from_request()}, handled=handled
    )
    return ori_handle_exception(self, exception)


WebRequest._handle_exception = _handle_exception


odoo_elasticapm/ir_cron.py
# -*- coding: utf-8 -*-
# Copyright 2019 Akretion (http://www.akretion.com).
# @author Sébastien BEAU <sebastien.beau@akretion.com>
# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).

from .base import elastic_apm_client, elasticapm, version_older_then

try:
    from odoo.addons.base.models.ir_cron import ir_cron as IrCron
    from odoo import api
except ImportError:
    try:
        from odoo.addons.base.ir.ir_cron import ir_cron as IrCron
        from odoo import api
    except ImportError:
        from openerp.addons.base.ir.ir_cron import ir_cron as IrCron
        from openerp import api


def before_cron(job):
    elastic_apm_client.begin_transaction("cron")
    elasticapm.set_user_context(user_id=job["user_id"])


def after_cron(job):
    if "name" in job:
        name = job["name"]
    else:
        name = job["cron_name"]
    elastic_apm_client.end_transaction(name)


ori_process_job = IrCron._process_job


if version_older_then("10.0"):

    def _process_job(self, job_cr, job, cron_cr):
        before_cron(job)
        ori_process_job(self, job_cr, job, cron_cr)
        after_cron(job)


else:

    @classmethod
    def _process_job(cls, job_cr, job, cron_cr):
        before_cron(job)
        ori_process_job(job_cr, job, cron_cr)
        after_cron(job)


IrCron._process_job = _process_job

ori_handle_callback_exception = IrCron._handle_callback_exception


def capture_exception(name, job_exception):
    elasticapm.label(
        exception_source="cron",
        exception_cron=name,
        exception_type=type(job_exception).__name__,
        exception_handled=False,
    )
    elastic_apm_client.capture_exception(handled=False)


if version_older_then("10.0"):

    def _handle_callback_exception(
        self, cr, uid, model_name, method_name, args, job_id, job_exception
    ):
        ori_handle_callback_exception(
            self, cr, uid, model_name, method_name, args, job_id, job_exception
        )
        job = self.browse(cr, uid, job_id)
        capture_exception(job.name, job_exception)


elif version_older_then("11.0"):

    @api.model
    def _handle_callback_exception(
        self, model_name, method_name, args, job_id, job_exception
    ):
        ori_handle_callback_exception(
            self, model_name, method_name, args, job_id, job_exception
        )
        job = self.browse(job_id)
        capture_exception(job.name, job_exception)


else:

    @api.model
    def _handle_callback_exception(
        self, cron_name, server_action_id, job_id, job_exception
    ):
        ori_handle_callback_exception(
            self, cron_name, server_action_id, job_id, job_exception
        )
        job = self.browse(job_id)
        capture_exception(job.name, job_exception)


IrCron._handle_callback_exception = _handle_callback_exception


odoo_elasticapm/ir_http.py
# -*- coding: utf-8 -*-
# Copyright 2019 Akretion (http://www.akretion.com).
# @author Sébastien BEAU <sebastien.beau@akretion.com>
# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).

from .base import elastic_apm_client, elasticapm, version_older_then
from .http import get_data_from_request

try:
    from odoo.addons.base.models.ir_http import IrHttp
    from odoo.http import request
except ImportError:
    try:
        from odoo.addons.base.ir.ir_http import IrHttp
        from odoo.http import request
    except ImportError:
        from openerp.addons.base.ir.ir_http import ir_http as IrHttp
        from openerp.http import request

SKIP_PATH = ["/connector/runjob", "/longpolling/", "/web_editor"]


ori_dispatch = IrHttp._dispatch


def skip_tracing():
    path_info = request.httprequest.environ.get("PATH_INFO")
    for path in SKIP_PATH:
        if path_info.startswith(path):
            return True
    return False


def before_dispatch():
    elastic_apm_client.begin_transaction("request")
    elasticapm.set_user_context(user_id=request.session.uid)


def after_dispatch(response):
    path_info = request.httprequest.environ.get("PATH_INFO")
    name = path_info
    for key in ["model", "method", "signal"]:
        val = request.params.get(key)
        if val and val not in name:
            name += " {}: {}".format(key, val)
    elasticapm.set_context(lambda: get_data_from_request(), "request")
    try:
        code = response.status_code
    except Exception:
        try:
            code = response.code
        except Exception:
            code = "NoCodeFound"
    elastic_apm_client.end_transaction(name, code)


if version_older_then("10.0"):

    def _dispatch(self):
        if skip_tracing():
            return ori_dispatch(self)
        else:
            before_dispatch()
            response = ori_dispatch(self)
            after_dispatch(response)
            return response


else:

    @classmethod
    def _dispatch(cls):
        if skip_tracing():
            return ori_dispatch()
        else:
            before_dispatch()
            response = ori_dispatch()
            after_dispatch(response)
            return response


IrHttp._dispatch = _dispatch


odoo_elasticapm/models.py
# -*- coding: utf-8 -*-
# Copyright 2019 Akretion (http://www.akretion.com).
# @author Sébastien BEAU <sebastien.beau@akretion.com>
# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).


from .base import elasticapm, version_older_then

try:
    from odoo import api, models
except ImportError:
    from openerp import api, models


def build_params(self, method):
    return {
        "name": "ORM {} {}".format(self._name, method),
        "span_type": "odoo",
        "span_subtype": "orm",
        "extra": {
            "odoo": {
                "class": self._name,
                "method": method,
                "nbr_record": hasattr(self, "_ids") and len(self) or 0,
            }
        },
    }


BaseModel = models.BaseModel
ori_create = BaseModel.create
ori_write = BaseModel.write
ori_search = BaseModel._search
ori_unlink = BaseModel.unlink


def write(self, vals):
    with elasticapm.capture_span(**build_params(self, "write")):
        return ori_write(self, vals)


@api.returns("self", lambda value: value.id)
def create(self, vals):
    with elasticapm.capture_span(**build_params(self, "create")):
        return ori_create(self, vals)


def _search(self, *args, **kwargs):
    with elasticapm.capture_span(**build_params(self, "search")):
        return ori_search(self, *args, **kwargs)


def unlink(self):
    with elasticapm.capture_span(**build_params(self, "unlink")):
        return ori_unlink(self)


if version_older_then("13.0"):
    unlink = api.multi(unlink)
    write = api.multi(write)

if version_older_then("12.0"):
    create = api.model(create)
else:
    create = api.model_create_multi(create)


if version_older_then("10.0"):
    _search = api.cr_uid_context(_search)
else:
    _search = api.model(_search)

BaseModel.create = create
BaseModel.write = write
BaseModel._search = _search
BaseModel.unlink = unlink
